domain/application/state/storage.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+
+	"github.com/juju/juju/domain/application"
+	"github.com/juju/juju/internal/errors"
+)
+
+// insertStorage constructs inserts storage directive records for the application.
+func (st *State) insertStorage(ctx context.Context, tx *sqlair.TX, appDetails applicationDetails, appStorage []application.AddApplicationStorageArg) error {
+	if len(appStorage) == 0 {
+		return nil
+	}
+
+	// This check is here until we rework all of the AddApplication logic to
+	// run in a single transaction. There's a TO-DO in the AddApplication service method.
+	queryStmt, err := st.Prepare(`
+SELECT &charmStorage.name FROM charm_storage
+WHERE charm_uuid = $applicationDetails.charm_uuid
+`, appDetails, charmStorage{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	var storageMetadata []charmStorage
+	err = tx.Query(ctx, queryStmt, appDetails).GetAll(&storageMetadata)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return errors.Errorf("querying supported charm storage: %w", err)
+	}
+	supportedStorage := set.NewStrings()
+	for _, stor := range storageMetadata {
+		supportedStorage.Add(stor.Name)
+	}
+	wantStorage := set.NewStrings()
+	for _, stor := range appStorage {
+		wantStorage.Add(stor.Name)
+	}
+	unsupportedStorage := wantStorage.Difference(supportedStorage)
```

Comment:
This is left-aligned, so if you have the inverse, it won't error out, I don't know if this a problem or not?

Reply:
Nah, we just want to know if something is not included in the supported list.